Alpaka generic reduction

alexstrel opened the pull request:
#### PR description:

This is test PR, introducing generic alpaka blas ("transform-reduce") kernels
#### PR validation:
TBD

mmusich committed:
move SiPixelRechits{Tracks,Vertices} SoA monitoring for alpaka into the pixelgpu_dqm_sourceclient-live_cfg.py

fwyzard committed:
Fix BuildFiles, includes and comments

alexstrel committed:
added multi-src blas kernels

alexstrel committed:
a number of compilation fixes

alexstrel committed:
fixed validation part (accessors to alpaka views)

alexstrel committed:
minor cleaning, created new file

alexstrel committed:
corrected build script

alexstrel committed:
removed artifacts, fixed default warp sizes , added one method to initialize vector array

makortel commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -89,26 +95,49 @@ namespace cms::alpakatools {
+    inline constexpr T const *data()                      const { return m_data;            }
+    inline constexpr void resize(std::int32_t size)             { m_size = size;            }
+    inline constexpr bool empty()                         const { return 0 == m_size;       }
+    inline constexpr bool full()                          const { return maxSize == m_size; }
```
These don't look like the code formatting would have been applied.

alexstrel replied:
addressed in 70d18caa76fdd1e1f3221c48f10efc1aaf3c4448

makortel commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -9,15 +9,21 @@
 namespace cms::alpakatools {
 
-  template <class T, int maxSize>
+  template <class T, std::int32_t maxSize>
```
Just curious, what motivates the change from `int` to `std::int32_t`?

alexstrel replied:
addressed in 83a3a7f495b08a7673d5ef9301b930cc19111e64

makortel commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -89,26 +95,49 @@ namespace cms::alpakatools {
+  struct is_vector_array<cms::alpakatools::VecArray<T, N>> : std::true_type {};
+
+  template<typename T>
+  inline constexpr bool is_vector_array_v = is_vector_array<T>::value;
```
I'd call this `is_VecArray_v` (and I see now why `std::is_same_v` wouldn't really work).

alexstrel replied:
addressed in 9157c25514af21fb6bf23ab8a94ae32edcef2dd9

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaAtomicHelper.h:
We usually name the files according to the thing it defines, so this would be `atomic_type.h` (same comment for the other headers in this directory).

alexstrel replied:
addressed in 8dc7009819ae554780aeffbd75daedb3d5ca1895

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaAtomicHelper.h:
```diff
@@ -0,0 +1,37 @@
+
+namespace cms::alpakatools {
+
+  template <typename T, typename SFINAE = void> struct atomic_type;
```
Could you add documentation here about the purpose of this metafunction?

alexstrel replied:
addressed in 8dc7009819ae554780aeffbd75daedb3d5ca1895

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+#include "HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReducer.h"
+
+namespace cms::alpakatools{
+  namespace blas{
```
Could be
```suggestion
namespace cms::alpakatools::blas {
```

alexstrel replied:
I suggest to keep it as is for now, usually I consider chaining less convenient since one may want to clearly distinguish between different namespaces : say, I may define something outside blas namespace in the future. (Of course one could just create another namespace scope, but for me it'd look less cleaner, though it could be a matter of taste)

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+        //
+        using reduce_t       = reduce_t_;
+        using transformer_t  = Transformer;
+        using reducer_t      = Reducer;
```
Given CMS' naming convention (type names start with capital letter), these all lower-case type aliases will be confusing in a a non-trivial piece of code.

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+        TransformReduceFunctor(transformer_t t, reducer_t r) : transformer(t), reducer(r) {}
+
+        template< typename TAcc >
+        ALPAKA_FN_ACC inline reduce_t operator()(TAcc const& acc, const reduce_t &x, const reduce_t &y ) const { return reducer(x, y); }
```
Why does this function depend on `TAcc`?

(also, although minor, explicit `inline` is unnecessary in class member functions)

alexstrel replied:
Removed TAcc and inline (agree, not needed)

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+
+        ALPAKA_FN_ACC inline auto init()   const { return reducer.init(); }
+
+        ALPAKA_FN_ACC inline auto get_reducer()    const { return reducer; }
```
Does this function return a copy of `reducer` intentionally?

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+        using reducer_t      = Reducer;
+        //
+        transformer_t transformer;
+        reducer_t     reducer;
```
This class looks complex-enough that it would be good to follow the data member naming convention (e.g. a trailing underscore, `reducer_`).

Can the data members be made `private`?

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+  namespace blas{
+
+    template <typename reduce_t_, typename Transformer, typename Reducer = reduce::plus<reduce_t_>, bool site_unroll_ = false>
+    class TransformReduceFunctor {
```
This class template would benefit from short documentation of the template arguments and intended use.

Same comment for the other class templates in this file.

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+	  cms::alpakatools::VecArray<T*, nSrc> result;
+
+          for (int i = 0; i < nSrc; ++i) {
+            result[i] = const_cast<T*> (vec[i].data());
```
Why is a `const_cast` needed? If its only because `vec` is assumed to be `VecArray` whose `data()` member function is `const`, please add non-`const` `data()` overload.

makortel commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -89,26 +95,49 @@ namespace cms::alpakatools {
+    }
+
+    VecArray<T, maxSize> &operator=(const VecArray<T, maxSize> &) = default;
+    VecArray<T, maxSize> &operator=(VecArray<T, maxSize> &&)      = default;    
```
I'd find it clearest to have the constructors and assignment operators defined as the first functions of the class.

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
I think the `alpaka` prefix could be removed from the file, i.e. rename to `blasCore.h`.

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+
+        const   cms::alpakatools::VecArray<Txz*, nSrc> x;
+        mutable cms::alpakatools::VecArray<Tyw*, nSrc> y;
+        mutable cms::alpakatools::VecArray<Tyw*, nSrc> w;
```
Why are these two `mutable`?

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+        , reducer_params(params) { }
+
+        template< typename TQueue >
+        auto fetch_data(TQueue queue) const { reducer_params.fetch_reduce_data(queue); }
```
```suggestion
        auto fetch_data(TQueue& queue) const { reducer_params.fetch_reduce_data(queue); }
```
assuming `TQueue` is an alpaka Queue (same comment for all functions that take `TQueue` by value).

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+        //
+        count_t* count;
+        //
+        reducer_params_t &reducer_params;
```
Can the data members be made private?

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+        MultiSrcTransformReduceKernel(TransformReducer_t f, const Args& args ) : f(f), args(args) {}
+	    
+        template< typename TQueue >  
+        auto fetch(TQueue queue) const { args.template fetch_data<TQueue>(queue); }	
```
```suggestion
        auto fetch(TQueue& queue) const { args.template fetch_data<TQueue>(queue); }	
```

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h:
```diff
@@ -0,0 +1,277 @@
+
+      auto const nsrc = x.size();
+
+      if( nsrc != nSrc ) std::cout << "Incorrect number of sources\n" << std::endl;
```
It would be better to throw an exception

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h:
```diff
@@ -0,0 +1,72 @@
+        : result_h(alpaka::allocBuf<system_atomic_t, Idx>(alpaka::getDevByIdx(alpaka::PlatformCpu{}, 0), nSrc))
+        , result_d(alpaka::allocBuf<system_atomic_t, Idx>(devAcc, nSrc))
+        , partial(alpaka::allocBuf<device_atomic_t, Idx>(devAcc, n_blocks))
+        , count(alpaka::allocBuf<count_t, Idx>(devAcc, nSrc)) {
```
Memory allocations should be done with the functions defined in `HeterogeneousCore/AlpakaInterface/interface/memory.h`, see https://github.com/cms-sw/cmssw/tree/master/HeterogeneousCore/AlpakaInterface#allocating-queue-ordered-host-buffers-in-device-mapped-memory and https://github.com/cms-sw/cmssw/tree/master/HeterogeneousCore/AlpakaInterface#allocating-queue-ordered-device-buffers

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h:
```diff
@@ -0,0 +1,72 @@
+  namespace reduce {
+
+
+    template<typename T, typename TAcc, typename TDevAcc, typename TQueue>
```
What exactly are `TAcc` and `TDevAcc`?

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h:
```diff
@@ -0,0 +1,72 @@
+
+        using TBuf1DAcc     = alpaka::Buf<TAcc, device_atomic_t, Dim1D, Idx>;
+
+        using TCountBufAcc  = alpaka::Buf<TAcc, count_t, Dim1D, Idx>;
```
I'd use the type aliases from `HeterogeneousCore/AlpakaInterface/interface/memory.h`
```suggestion
        using TBufHost      = host_buffer<system_atomic_t[]>;
        using TBufAcc       = device_buffer<TAcc, system_atomic_t[]>;

        using TBuf1DAcc     = device_buffer<TAcc, device_atomic_t[]>;

        using TCountBufAcc  = device_buffer<TAcc, count_t[]>;
```

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h:
```diff
@@ -0,0 +1,72 @@
+
+        //static ReducerResource reduction_buffers; 
+        static ReducerResource<T, TAcc, TDevAcc, TQueue>& get_reduction_resources(const TDevAcc &devAcc, TQueue queue, Idx nSrc, Idx n_blocks) {
+          static ReducerResource<T, TAcc, TDevAcc, TQueue> instance(devAcc, queue, nSrc, n_blocks);
```
Is there any reason why this function-`static` variable could not be `const`?

Actually, the set of constructor arguments is very suspicious for a variable that is effectively a global.

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReducer.h:
```diff
@@ -0,0 +1,45 @@
+      return v;
+    }
+
+    template <typename T, typename U> constexpr std::enable_if_t<std::is_arithmetic_v<T> and std::is_arithmetic_v<U>, T> set(U x) { return static_cast<T>(x); }
```
The purpose of this function is not clear to me

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaTransformer.h:
```diff
@@ -0,0 +1,47 @@
+namespace cms::alpakatools {
+  namespace transform {
+
+    template <typename T> struct xpy {
```
This nomenclature is unclear to most in CMS

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+    exit(EXIT_FAILURE);
+  }
+  
+  static const Idx N     = 1 << 8;
```
Could be `constexpr`.

Do these need to be `static`?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+    // Select specific devices
+    auto const platformHost = alpaka::PlatformCpu{};
+    auto const devHost      = alpaka::getDevByIdx(platformHost, 0);
+    auto const platformAcc  = alpaka::Platform<Acc>{};    
```
Where is `Acc` type (alias) defined?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+    if ( alpaka::getAccName<Acc>() == "AccCpuSerial<3,unsigned int>" ) {
+      std::cout << "... skipped" << std::endl;	    
+      return EXIT_SUCCESS;
+    }    
```
What is the purpose of this conditional?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+
+    // Select queue
+    using QueueProperty = alpaka::NonBlocking;
+    using QueueAcc      = alpaka::Queue<Acc, QueueProperty>;
```
These types should be taken from the `ALPAKA_ACCELERATOR_NAMESPACE`, i.e. you can replace all uses of `QueueAcc` with `Queue`. See e.g.
https://github.com/alexstrel/cmssw/blob/b4b942d08892dab4db669ce73f70dc104fe456f8/HeterogeneousCore/AlpakaInterface/interface/config.h#L46-L49

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+    std::cout << "Test nsrc kernels on " << alpaka::getName(device) << '\n';
+    
+    // Select specific devices
+    auto const platformHost = alpaka::PlatformCpu{};
+    auto const devHost      = alpaka::getDevByIdx(platformHost, 0);
```
Use these functions instead
https://github.com/alexstrel/cmssw/blob/b4b942d08892dab4db669ce73f70dc104fe456f8/HeterogeneousCore/AlpakaInterface/interface/host.h#L10-L14
```suggestion
    auto const& platformHost = cms::alpakatools::host_platform();
    auto const& devHost      = cms::alpakatools::host();
```

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+    auto const devHost      = alpaka::getDevByIdx(platformHost, 0);
+    auto const platformAcc  = alpaka::Platform<Acc>{};    
+    
+    auto const computeQueue = Queue(device);    
```
IIRC `const` `Queue` wouldn't work with Alpaka API. Has this changed?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+        
+    // Use increasing values as input
+    //
+    HostViewType xView = alpaka::createView(devHost, xv.data(), extent);//create 3D view 
```
How is this view 3D when `extent` is 2D?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+                      << gres << std::endl;       
+    }
+       
+    alpaka::wait(computeQueue);  
```
Why is this `wait()` needed?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+     
+    alpaka::memcpy(computeQueue, check_yView, yAcc[nSrc-1], extent);    
+
+    msrc_axpyNorm_functor.template fetch<QueueAcc>(computeQueue);    
```
Is the `template ` needed here (we are not in a function template)?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+        );
+
+    // Copy device -> host
+    alpaka::wait(computeQueue);
```
Why is this `wait()` needed?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+       
+    alpaka::wait(computeQueue);  
+      
+    // Simulate
```
What is being simulated here?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+    std::cout << "Multisrc reduction kernel.\n";
+    std::cout << "Vector Size:" << N << "x" << ", src number:" << nSrc << "\n";
+
+    std::cout << "Sampled result checks are correct!\n";
```
How is the correctness verified? Ideally the test program would return with non-zero exit code in case the results are incorrect.

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+
+    return res;
+  }
+  // not needed...
```
Can it be removed then?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+          uint64_t nSrc = 1,
+          bool site_unroll_ = false 
+          >
+struct AxpyNorm : public blas::TransformReduceFunctor<reduce_t, Transformer, Reducer, site_unroll_> {
```
I think this functor would benefit from some documentation describing what it is doing.

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+
+  const DataType a;  
+  
+  AxpyNorm(const DataType &a, const DataType &, transformer_t t, reducer_t r) : blas::TransformReduceFunctor<reduce_t, Transformer, Reducer, site_unroll_>(t, r), a(a) {} 
```
Why is the second `DataType` argument ignored?

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+  const DataType a;  
+  
+  AxpyNorm(const DataType &a, const DataType &, transformer_t t, reducer_t r) : blas::TransformReduceFunctor<reduce_t, Transformer, Reducer, site_unroll_>(t, r), a(a) {} 
+  AxpyNorm(const DataType &a, const DataType &) 
+        : blas::TransformReduceFunctor<reduce_t, Transformer, Reducer, site_unroll_>(Transformer(), Reducer()) 
+        , a(a) 
+        {}      
```
I believe this could be
```suggestion
  AxpyNorm(const DataType &a, const DataType &b) : AxpyNorm(a, b, Transformer(), Reducer()) {}
```

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReducer.h:
```diff
@@ -0,0 +1,45 @@
+    };
+
+    template <typename T, typename U>
+    constexpr auto get_alpaka_reducer(const cms::alpakatools::reduce::plus<U> &) { return cms::alpakatools::reduce::plus<T>();     }
```
The purpose of this function is not clear to me.

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
```diff
@@ -0,0 +1,305 @@
+                                                                 func_t,
+                                                                 nSrc, false > (devAcc, computeQueue, a, a, xAcc, yAcc, yAcc, xAcc);
+
+    AxpyNormProductKernel<Acc, decltype(devAcc), decltype(msrc_axpyNorm_functor)> axpyNormProduct;
```
We have accustomed in a nomenclature where a "kernel" is the device function that is called from the host (i.e. `__global__` in CUDA, or the functor passed to `alpaka::exec()`). IIUC, here `axpyNormProduct` is the "kernel", whereas `msrc_axpyNorm_functor` is a device-side function/functor. Creating such functors with a function that has `Kernel` in its name is a bit confusing.

alexstrel replied:
addressed in 8dc7009819ae554780aeffbd75daedb3d5ca1895

makortel commented on HeterogeneousCore/AlpakaInterface/test/alpaka/testReducer.dev.cc:
I see this test as a useful for the entire transform-reduce infrastructure, but I feel it is too complicated to serve as an example for a case like the primary vertexing, where the reducer would be called from a user-written kernel that is otherwise independent of this transform-reduce infrastructure.

Could you create another test whose structure would be closer to the structure of the primary vertexing case?

makortel commented on HeterogeneousCore/AlpakaInterface/interface/VecArray.h:
```diff
@@ -89,26 +95,49 @@ namespace cms::alpakatools {
+    VecArray(const VecArray<T, maxSize> &) = default;
+    VecArray(VecArray<T, maxSize> &&)      = default;
+
+    ALPAKA_FN_ACC VecArray(const T& value) {
```
I find this constructor overload a bit confusing. It initializes the memory up to the capacity of the `VecArray`, but does not change its size. What is its purpose?

alexstrel replied:
This constructor treated VecArray a-la std::array (with compile-time size), fixed in 790dc200051a0cfc10b8963291880f9acbb95b95

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h:
```diff
@@ -0,0 +1,72 @@
+      private: 
+        TBufHost      result_h;
+        TBufAcc       result_d;
+        TBuf1DAcc     partial;
```
Do I understand correctly that these owning buffer objects get copied to the device (as part of the arguments to `alpaka::exec()`, as is done in the test)?

If I understood correctly, I'd look into stronger separation of objects owning the data (that is really a host-only property) from the algorithm that accesses the data (via a "view", minimally a pointer). The benefits would include copying less data as part of the kernel launch, no chance to accidentally access host memory from the device code, and the memory access pattern would match to what we do elsewhere.

alexstrel replied:
No, if I understood your point.  Let me explain this part:  purpose of ReducerResource class is to manage a single static instance of it (via get_reduction_resources) for a given set of parameters, see 
https://github.com/alexstrel/cmssw/blob/Alpaka-generic-reduction/HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h#L264
What is used in particular algorithms (or, better, compute kernels) is its view via TransformReduceArgs instance: 
https://github.com/alexstrel/cmssw/blob/Alpaka-generic-reduction/HeterogeneousCore/AlpakaInterface/interface/blas/alpakaBlasCore.h#L267

makortel replied:
Thanks for the explanation. Unfortunately static or global storage is highly undesirable because of it effectively disallows concurrent use of these functions when the `ReducerResource<...>` type is the same.

alexstrel replied:
fixed in 46031f0e18b045db84d9439952e831b1cf82d627

makortel commented on HeterogeneousCore/AlpakaInterface/interface/blas/alpakaReduceResource.h:
```diff
@@ -0,0 +1,72 @@
+
+        auto get_count_ptr()        { return count.data(); }
+        // 
+        auto fetch_reduce_data(TQueue queue) { alpaka::memcpy(queue, result_h, result_d); }
```
An alternative pattern that would allow allocating `result_h` in the case where the reduction result is not needed on the host (and could also help separating the host-side functionality from device-side functionality)  would be to allocate `result_h` in the function body, and return it to the caller after the `memcpy()` call. I'd then rename the function to something along `copyResultToHost()`.

Eventually it would be good to figure out a way that would avoid this `memcpy()` for CPU backend(s).

makortel reviewed (commented):
First pass comments, focusing mostly on the overall structure (I trust you on the algorithm logic).

alexstrel committed:
applied clang-format to all new files

alexstrel committed:
adopted standard notations

alexstrel committed:
replaced vector_array with VecArray

alexstrel committed:
changed file/class/structure names, added comments

alexstrel committed:
fixed array size

alexstrel committed:
added template parameter to avoid unnecessary initialization of the host pointer

alexstrel committed:
blas resources now thread local

alexstrel committed:
updated test application (missed in the previous commit)

alexstrel committed:
minor changes in method defitions (removed unused args etc.)

alexstrel committed:
a number of fixes (typos)

alexstrel committed:
a few changes to fix hip backend, still issues due to static assert (argument being not trivially copyable)

mmasciov committed:
Add masking for phase-2

mmasciov committed:
Add cut-based filter before output conversion

rovere committed:
Implement Phase 2 PixelTracks CA Extension

- Enable OT SoA producers, increase SoA size for pixelTracks.
- Update CAHitNtuplet to support Phase2 OT CA extension.
- Update also the configuration parameters to include the 3 additional
barrel layers from the OT.
- Correctly compute the index used to address rotations.
The index in the SiStipHitSoA to address the proper rotation of the
module is now correctly computed by couting all the preceeding pixel
modules and simply incrementing that number by one, for every new
P-module in the OT barrel.
The code and logic is extremely fragile and relies on a couple of
assumptions:
- all pixel modules will be used in the CA, not matter what.
- the order of modules in Phase2OTRecHitsSoAConverter and in
CAHitNtuplet is the same.
- Add producer to merge Pixel and tracker recHits
- Modify PixelTrack converter (SoA -> legacy) to be able to handle the extension to OT in Phase-2
- Update hard-coded constants to be able to run CA+Extension.
- Add new TrackerTraits topology
- Fix CA to work both with and without OT extension
The traits guarding Phase2 and Phase2OT should now be guarded correctly

Co-Authored-By: Adriano Di Florio <adriano.di.florio@cern.ch>
Co-Authored-By: Marco Musich <marco.musich@cern.ch>
Co-Authored-By: Jan Schulz <jan.gerrit.schulz@cern.ch>
Co-Authored-By: Luca Ferragina <ferragina.luca@gmail.com>

mmusich committed:
allow SingleMuPt15Eta0p_0p4 configuration fragment for NGT scouting workflow

felicepantaleo committed:
Define MultiCollections in RecHitMapProducer

rovere committed:
Cache PinPSinOTBarrel. Use vector and unordered_map instead of set.

rovere committed:
Move geometry-related logic in BeginRun.

JanGerritSchulz committed:
Update logic in PixelTrackProducerFromSoAAlpaka to reuse the HitModuleStart for the OT RecHits as well

JanGerritSchulz committed:
Fix a few CA bugs, add procModifier to enable the extension

Co-Authored-By: Marco Musich <marco.musich@cern.ch>
Co-Authored-By: Marco Rovere <marco.rovere@cern.ch>
Co-Authored-By: Luca Ferragina <ferragina.luca@gmail.com>

mmusich committed:
Restructure Phase2 PixelTracking configuration and ngtScouting modifier
interaction with LST

- introduce MVA-based high purity selection for extended pixelTracks
- introduce PixelVertexing sequence
- include this PixelVertexing sequence in combined PixelTracksAndVertices sequence
- make high purity selected PixelTracks the final PixelTracks for CAExtension
- optmise CA memory usage
- implement layer-pair dependent ptCuts
- compute minRadius once per layer pair
- tigthen CA parameters
- split layer pair (2,28) in three parts

- Enable OT-trimmed seeding for LST when trackingLST and CAExtension are used
- ngtScouting without trackingLST means generalTracks==pixelTracksHP
- ngtScouting with trackingLST means generalTracks==(pixelTracksHP + T5s)

Co-Authored-By: Jan Schulz <jan.gerrit.schulz@cern.ch>
Co-Authored-By: Marco Rovere <marco.rovere@cern.ch>
Co-Authored-By: Luca Ferragina <ferragina.luca@gmail.com>

Commit:
Added SoABlocks feature

Parsifal-2045 committed:
Reduce code duplication and improve readability in SoA merging

Parsifal-2045 committed:
Add column types to SoA descriptor

Parsifal-2045 committed:
Fix check-by-index for scalar in recHits SoA

amecca committed:
[TkAl][PixBary][FIX] Do not clobber output of multiple PixBary extract jobs

Add to the output directory path the names of: the single job, the
alignment and the IOV (as it is done for the working directory path).
This avoids overwriting the output when extracting from multiple alignments
and/or IOVs simultaneously.

JanGerritSchulz committed:
Add option in PixelTrackProducerFromSoAAlpaka to exclude quadruplets with skipping layers

JanGerritSchulz committed:
Add new vector cuts for doublet building in CA

JanGerritSchulz committed:
make chi2 cut on CA Tracks dependent on the number of ReCHits

JanGerritSchulz committed:
change default for Phase2 CA extension to requiring quadruplets to have no skipped layers

JanGerritSchulz committed:
Apply improved CA cuts for regular and extend configurations

JanGerritSchulz committed:
Change effect of useCAExtension flag in PixelTrack converter to include (discard) OT hits if enabled (disabled)

JanGerritSchulz committed:
Unify CA cuts for min/max inner/outer r/z based on the type of layer (endcap/barrel)

Parsifal-2045 committed:
Remove OT recHits conversion from sequences without CA extension

alexstrel committed:
fixed bugs in VecArray constructors, fixed hip compilation

alexstrel committed:
removed const

JanGerritSchulz committed:
Fix isOuterLadder of CA PixelDoubletAlgo for Phase-2

JanGerritSchulz committed:
Make Patatrack Pixel Tracking the default for Phase2

Parsifal-2045 committed:
Add CA Extension workflows to ph2_hlt short matrix and IB matrix

Parsifal-2045 committed:
Simplify isPh2Pixel check

Move debug printouts to LogDebug and silence compiler warnings

JanGerritSchulz committed:
Add customization of  HLT Run3 for new CA cuts introduced in CA extension

- after the customization, the cuts action in Run-3 should be unchanged compared to the previous implementation
- newly intorduced cuts are either not used or configured in such a way that they replicate removed/renamed cuts

Fix size of default ptCut vector in CA for Phase1

Adjust customization functions for Alpaka Pixel Tracking

- fix name of function in customizeHLTforCMSSW to match the PR number
- fix customizePixelTracksForTriplets for CA extension
- fix ptCut overwriting in offline Alpaka Pixel Tracks

mmusich committed:
restructure HLTPhase2PixelTracksAndVerticesSequence for vertex trimming process modifier

JanGerritSchulz committed:
Improve PixelTrack selection settings for non-extended CA in Phase-2

JanGerritSchulz committed:
Remove printouts and cleanup PixelTrack converter from SoA

JanGerritSchulz committed:
Rename maxChi2Quadruplets in Pixel Tracking to maxChi2TripletsOrQuadruplets

Parsifal-2045 committed:
Remove unecessary loop and alpaka::memcpy

mmusich committed:
Add an heterogeneous only ph-2 upgrade path and menu

Contains all modules that either can be offloaded to GPU or the necessary inputs.
No conversions back to host.

- support a version that runs also the CAExtension
- support a version that runs also LST
- vertexing not yet default

JanGerritSchulz committed:
Change back max pT for multiple scattering in BrokenLinFit

JanGerritSchulz committed:
Set max number of modules in ClusteringConstants to the actual number of IT modules in the Phase2 (D110) geometry

JanGerritSchulz committed:
Improve CA and SimplePixelTopology

- fully remove cellPtCut and isBarrel from CA configuration
- clean and restructure SimplePixelTopology
- restructure CA fillDescriptions to read the parameters from the TrackerTraits defined in SimplePixelTopology
- unify and simplify CA setup read from geometry
- remove isBarrel from hltPhase2PixelTracksSoA config
- fix filling of ptCut vector for Pixel Triplets customization

JanGerritSchulz committed:
Introduce phase2LegacyPixelTracks procModifier to enable the usage of the formerly default Legacy Pixel Tracks in HLT Phase-2

JanGerritSchulz committed:
Update SoA view accessors in CA OT RecHit converter chain

Parsifal-2045 committed:
Simplify usage of +1 in merging hitModuleStart SoAs

JanGerritSchulz committed:
Improve function getNSkippedLayer in PixelTrackProducerFromSoAAlpaka

mmusich committed:
addition of offline Patatrack workflow for Phase2 CA extension with OT

Parsifal-2045 committed:
Reset modules' maps and indices in beginRun

Martin-Grunewald committed:
Move some more scripts to 15_1

Martin-Grunewald committed:
Integration of configuration changes from #49147

- remove customization of menus for #49147

mmusich committed:
add phase2CAExtension to the list of modifiers used to run HLT OT RecHits validation on demand

mmusich committed:
HLT menu development for 15_1_X (branch for 16_0_X)

These updates correspond to the "V1.0" HIon menu for the 2025 PbPb data-taking period.
This update includes the integration of the following JIRA tickets:

- [CMSHLT-3147](https://its.cern.ch/jira/browse/CMSHLT-3147) [TRK,HIN] Remove DQM plugins from DQM_PixelReconstruction_v and DQM_HIPixelReconstruction_v
- [CMSHLT-3643](https://its.cern.ch/jira/browse/CMSHLT-3643) [DAQ,TSG] Update HLT to run with secondary TCDS
- [CMSHLT-3654](https://its.cern.ch/jira/browse/CMSHLT-3654) [TRK,HIN] HLT-Tracking updates for 2025 HI menu: switching on the good edge algo for pixel local reco
- [CMSHLT-3658](https://its.cern.ch/jira/browse/CMSHLT-3658) [HIN] HLT menu HIon for 2025 PbPb run
- [CMSHLT-3659](https://its.cern.ch/jira/browse/CMSHLT-3659) [TRK,HIN] RawSecond in HIon menu
- [CMSHLT-3666](https://its.cern.ch/jira/browse/CMSHLT-3666) [TRK,HIN] Clean-up instances of PixelTrackProducerFromSoAAlpaka* and SiPixelRecHitFromSoAAlpaka* in the HLT combined tables

AuroraPerego committed:
add rechits/simhits to dumper

AuroraPerego committed:
use one flags to save hits

Parsifal-2045 committed:
Let the framework handle host to device copies

Remove unecessary shifted vector of recHits per module

Squash

mmusich committed:
add also Heavy Ions overrides for online pixel GPU client

Parsifal-2045 committed:
Correctly handle skipped tracks when converting SoA to Legacy

cmsbuild committed:
Merge pull request #49226 from Dr15Jones/clangChange

Updated custom suffix syntax required by clang

cmsbuild committed:
Merge pull request #49222 from perrotta/cloneOfRun2DataGtIn150X

Replace the GT for Run2 data re-reco in autoCond with a 150X clone of previous 141X_dataRun2_v3

cmsbuild committed:
Merge pull request #49216 from AlexAlbert41/update-MDS-trigger-thresholds

updated L1 thresholds for MDS Trigger for CSCs for Era 2025F

cmsbuild committed:
Merge pull request #49212 from smuzaffar/reconstruction-llvm21-virtual-method-final-class

[RECONSTRUCTION] Fix for virtual method inside a final class

cmsbuild committed:
Merge pull request #49211 from smuzaffar/simulation-llvm21-virtual-method-final-class

[SIMULATION] To virtual method inside a final class; make classes non-final

cmsbuild committed:
Merge pull request #49195 from zhiyuan011/zhhuang-plot

Add a plotting script for BS offline

JanChyczynski committed:
LHCInfoPerLS endFill conflicting IOVs fix

Add warning comment that lumisection query filter is not always 'greater than'

fabiocos committed:
Squashed commit of the following:

commit f20ddceebf0fc84c195e0a1e3ce98d8aa8bd0760
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Tue Oct 28 10:53:47 2025 +0100

    additional cleaning and added a method in MTDDigiGeometryAnalyzer.cc

commit 8f3142b5329994471e3b09d6886ecd5d9b9c351a
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Mon Oct 27 15:09:08 2025 +0100

    cleaned the branch, should be ready for a PR

commit 145c2c4be0f3ebee1dd6134f2fc82dba7bfed125
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Tue Oct 21 15:21:10 2025 +0200

    etl v12 working, there is still a sanity check on the PPMTDParameters vectors to restore

commit c2ea4bcb256999d92a5524b9c39e7879153582c9
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Fri Oct 10 14:56:28 2025 +0200

    small modification to etl.xml v10 due to a small shift in x found for a module, now v10 and v12 are equivalent and both correct

commit 70621667e09ec9d720132ccb5d31106cad01a233
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Thu Oct 9 14:48:54 2025 +0200

    added the last missing piece to upgradeWorkflowComponents.py

commit d2a6b62978ca623d02fbdb9e48a2aa5b8191c7ba
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Thu Oct 9 11:36:27 2025 +0200

    almost generated Run4D201

commit c551afaef2e82c99df5838a2c7af64802cfd155d
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Wed Oct 8 17:27:57 2025 +0200

    added ETL v12 and introduced the scenario D201 in dictRun4Geometry.py and in the README.md

commit c928467901fe3cbb37383f07f1ed1ddde25957e6
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Mon Oct 6 15:08:23 2025 +0200

    changed default scenario, everything works for vv2 now

commit 4f50c6eca962002eac447fe99a49c3319817290c
Author: Leonardo Lanteri <leonardo.lanteri@gmail.com>
Date:   Fri Sep 26 17:04:30 2025 +0200

    small modifications, mainly cleaning, NB the Default scenario that was set

commit 712188b81ca6d0b122a0bfde54ae00fb31014ae5
Author: Fabio Cossutti <fabio.cossutti@ts.infn.it>
Date:   Thu Sep 18 15:13:51 2025 +0200

    Update orderETLSector to work without module number, update RecoMTD/DetLayers test for topology validation

commit 68570f031559761b8fecd7d1ed1458232b947ffe
Author: Fabio Cossutti <fabio.cossutti@ts.infn.it>
Date:   Wed Sep 17 15:22:52 2025 +0200

    Update verbosity for consistency among tests, update orderETLSector, fix version usage in topology

commit 11df80772da00d7e36a8dcf848529e52800b1a13
Author: Fabio Cossutti <fabio.cossutti@ts.infn.it>
Date:   Tue Sep 16 11:35:59 2025 +0200

    Add workflows for D200, update service hybrid methods, debugging statements (to be removed)

commit 50ab8aa78f896a572f0ca3b0d2fe5ae349bd6167
Author: Fabio Cossutti <fabio.cossutti@ts.infn.it>
Date:   Fri Sep 12 15:03:13 2025 +0200

    Add ETL v11, adapt code, clean existing implementation

commit aa189c877aca336fc711714a7fe63dc3661d1b2c
Author: Fabio Cossutti <fabio.cossutti@ts.infn.it>
Date:   Wed Sep 10 17:57:29 2025 +0200

    Updates for scenario v11

commit e265364b880b79f2cb034592fecf990d469283ad
Author: Fabio Cossutti <fabio.cossutti@ts.infn.it>
Date:   Wed Sep 10 17:39:30 2025 +0200

    Add a ETL v11/scenario D200, for now equal to v9

Co-authored-by: Leonardo Lanteri <leonardo.lanteri@gmail.com>

tomalin committed:
Changes requested by Nazar

tomalin committed:
Changes requested by Nazar

cmsbuild committed:
Merge pull request #49230 from fwyzard/from-CMSSW_16_0_X_2025-10-27-2300

Explicitly export CUDA kernels in shared libraries

fabiocos committed:
Restore .gitignore, Restore BTL printout

Commit:
code checked

mmusich committed:
Introduce change of input file number for GPUvsCPU tests to accomodate change of event content

- See [CMSHLT-3147](https://its.cern.ch/jira/browse/CMSHLT-3147) for more details.

Commit:
Add Run3_2025_NEON in Eras.py

cmsbuild committed:
Merge pull request #48854 from leobeltra/workaround#47931_removed

Removed workaround for illegal memory access

cmsbuild committed:
Merge pull request #49179 from pallabidas/EMTF_NN_hls

[L1T] replacing tensorflow with hls4ml model for EMTF NN

cmsbuild committed:
Merge pull request #49197 from konpas17/hcal-hb_llp_thresholds

HCAL: Move HB LLP bit energy thresholds to CondDB

cmsbuild committed:
Merge pull request #49240 from stahlleiton/Run3_2025_NEON_CMSSW_16_0_X

Add Run3_2025_NEON in Eras.py

cmsbuild committed:
Merge pull request #49205 from srimanob/160_MoveMiniCustomizeToEventContent

Remove miniAOD_customizeOutput, and define miniaod file parameter directly in EventContent

Commit:
Correct the documentation in HGCSiliconDetId

alaperto committed:
Cosmics Strip Clusters thresholds DQM

alaperto committed:
Update Disabled ROCs for Online DQM Pixel Summary

smuzaffar committed:
HcalObjRepresent: Fix for destroyed point after the full-expression

smuzaffar committed:
MaterialBudgetData: Return a const& to string instead of making a copy

Commit:
Update one of the test script which may help in automaziation of IsoTrack calibration

casarsa committed:
Disabled debugging output

Commit:
correct two rebasing errors

Dr15Jones committed:
Proper timing for post Modules init signal

Have postModulesInitializationFinialized signal happen before pre begin job.

Dr15Jones committed:
Add new initialization signals to Tracer

cmsbuild committed:
Merge pull request #48921 from cms-ngt-hlt/ngt_phase2CAExtension

[NGT] Extension of CA Pixel Tracking to Phase 2 Outer Tracker barrel

cmsbuild committed:
Merge pull request #49248 from smuzaffar/llvm21-warn-destroyed-point

HcalObjRepresent: Fix for destroyed point after the full-expression

cmsbuild committed:
Merge pull request #49220 from denizsun/hcal-hep07-sporadic-data

[HCAL-DQM] Mask channels in HEP07 (RM 3, RMfib 4) affected by sporadic bad data

cmsbuild committed:
Merge pull request #49250 from smuzaffar/llvm21-warn-destroyed-obj1

MaterialBudgetData: Return a const& to string instead of making a copy

cmsbuild committed:
Merge pull request #49243 from bsunanda/Phase2-hgx364H

Phase2-hgx364H Correct the documentation in HGCSiliconDetId

alexandertuna committed:
Add fake-or-duplicate track rate plots to LST standalone plots

Commit:
mkfit modifier and clean up

leonardolanteri committed:
fixed vector in MTDTopologyEP.cc

leonardolanteri committed:
ran code-format

Electricks94 committed:
Introduce MultiCollection as alias for vector<RefProd<T>>

Electricks94 committed:
Added better debug messages

Commit:
Try to avoid the issue concerning position to detId conversion

mandrenguyen committed:
Revert "Merge pull request #49179 from pallabidas/EMTF_NN_hls"

This reverts commit 3351dfbe81b4a215796f153cad979f19cd1f7035, reversing
changes made to 6a7bc96280b95909d390d3a840e6f4f6aabdf56c.

Commit:
Correct some of the tesing code in Geometry/HGCalGeometry

casarsa committed:
Improved code readability

leonardolanteri committed:
minor change in upgradeWorkflowComponents.py

cmsbuild committed:
Merge pull request #49263 from mandrenguyen/revert_PR49179

Revert "Merge pull request #49179 from pallabidas/EMTF_NN_hls"

Parsifal-2045 committed:
Get offsetBPIX2 from the geometry

- remove reliance on hard-coded value in simplePixelTopology
- support older geometries (tested from D95)

cmsbuild committed:
Merge pull request #49264 from bsunanda/Phase2-hgx364J

Phase2-hgx364J Correct some of the testing code in Geometry/HGCalGeometry

cmsbuild committed:
Merge pull request #49170 from gartung/gartung-PerTools-AllocMonitorAnalyze

PerfTools/ModuleAllocMonitor: Add source c++type to moduleAllocMonitor.log, sum all source activity for each event in source module

cmsbuild committed:
Merge pull request #49257 from Dr15Jones/moreTracerSignals

Added new signals to Tracer

cmsbuild committed:
Merge pull request #49262 from bsunanda/Phase2-hgx364I

Phase2-hgx364I Try to avoid the issue concerning position to detId conversion

cmsbuild committed:
Merge pull request #49258 from alexandertuna/AddPixelSeedChargeAndDeltaPhiPostLstOd

Add pixel charge and deltaphi to LST ntuple

cmsbuild committed:
Merge pull request #49042 from fwyzard/alpaka_2.0

Update HeterogeneousCore/AlpakaInterface for alpaka 2.0

smuzaffar committed:
MkFitCore/Track: Fix for llvm21 warning calling bzero on pointer to non-trivially copyable type

Electricks94 committed:
Changed MultiCollection to RefProfVector

hjbossi committed:
[FSC DQM]: Initial commit of FSC plus implementation in DQM.

Commit:
Try to fix the failures as repoed in the issue #48347

Commit:
Code check

cmsbuild committed:
Merge pull request #49272 from Parsifal-2045/fixOffsetBPIX2

[NGT] Add support for older geometries in PixelTracking CA

cmsbuild committed:
Merge pull request #49251 from smuzaffar/llvm21-fix-bzero-non-trivially-copyable

MkFitCore/Track: Fix for llvm21 warning calling bzero on pointer to non-trivially copyable type

cmsbuild committed:
Merge pull request #49269 from cms-AlCaDB/conflicting_iovs

LHCInfoPerLS endFill conflicting IOVs fix

cmsbuild committed:
Merge pull request #49221 from casarsa/mc-btlnewdigi_squash

Update of the BTL digitization model

Dr15Jones committed:
Pass new signals to Tracer file

- handle new startup related signals in Tracer file
- Update edmTracerCompactLogViewer to handle new file entries
- Update web view of timing for new signals.

fwyzard committed:
Add missing dependency on json

AdrianoDee committed:
Proper clamping for CA doublets

mmusich committed:
Update Phase 2 HLT timing test (post #48921)

- Add test to run the Phase 2 HLT timing menu forcing CPU backend
- modify the NGT menu test to use ngtScouting and phase2CAExtension modifiers

adzel-werk committed:
Add InputTag parameters to fillDescriptions method

Added new InputTag parameters for HepMC and RecoProtons.

adzel-werk committed:
Update InputTags for HepMC and RecoProtons

Commit:
Mitigate impact of FPIX hole in UPC reco

cmsbuild committed:
Merge pull request #49279 from mmusich/mm_update_runHLTTiming_post-48921

Update Phase 2 HLT timing test (post #48921)

cmsbuild committed:
Merge pull request #49275 from bsunanda/Phase2-hgx364K

Phase2-hgx364K Try to fix the failures as reported in the issue #48347

cmsbuild committed:
Merge pull request #49260 from alexandertuna/LstAddFakeOrDuplRatePlotsPostLstOd

Add fake-or-duplicate track rate plots to LST standalone plots

cmsbuild committed:
Merge pull request #49228 from tomalin/ianTrkTopoP2

Add function and enum aliases for P2 tracker topology with more obvious names

cmsbuild committed:
Merge pull request #49227 from stahlleiton/PbPbEWSkim_CMSSW_16_0_X

Add EW skim for 2025 PbPb

cmsbuild committed:
Merge pull request #49231 from bdanzi/mkFitModifierHLTPhase2

[mkFit] procModifier for Phase-2 HLT tracking and track candidates selections

cmsbuild committed:
Merge pull request #49237 from mmusich/pixel_dqm_client_changes_for_CMSHLT-3147

Pixel GPU Online DQM Client changes for CMSHLT-3147

Dr15Jones committed:
Switch to use RNTupleTemp* plugins

-Switch from RNTuple* to RNTupleTemp* as the latter more closely
behave like their Pool* counterparts.

cmsbuild committed:
Merge pull request #49284 from stahlleiton/UPCTrack_CMSSW_16_0_X

Mitigate impact of FPIX hole in UPC reco

cmsbuild committed:
Merge pull request #49253 from bsunanda/Run3-alca263

Run3-alca263 Update one of the test scripts, which may help in the automatization of IsoTrack calibration

Dr15Jones committed:
RNTuple requires MEtoEDM<long> have a dictionary

cmsbuild committed:
Merge pull request #49276 from Dr15Jones/tracerFileStartup

Pass new signals to Tracer file

cmsbuild committed:
Merge pull request #48969 from dan131riley/ttrecache-refactoring

PoolSource TTreeCache refactoring

cmsbuild committed:
Merge pull request #49278 from fwyzard/add_missing_json_dependency

Add missing dependencies on json

Commit:
disable ubsan checking for global summary fns

Commit:
add comment

mmusich committed:
update hltPathOwners.json for HLT menu HIon 2025 V1.0

mmusich committed:
bumped version of Paths for HLT HIon-2025 V1.0 (base: /dev/CMSSW_15_0_0/HLT/V129)

cmsbuild committed:
Merge pull request #49282 from AdrianoDee/nHitsToCells_clamp_fix_160X

Proper Clamping for CA Doublets Histogram

cmsbuild committed:
Merge pull request #49234 from leonardolanteri/ll-ETL_detId

MTD Geometry: Update ETL numbering scheme, add corresponding new geometry scenarios

Commit:
Try to remove fine scintillator tile dependent tests for layers with no fine tiles

Commit:
Code check

mmusich committed:
upgrade phase1_2025_realistic_hi autoCond and run3_mc_HIon key to use a 151X based Global Tag

smorovic committed:
use 64-bit pointers for buffer offsets which allows buffers larger than 4G.
Only DAQSource is adapted to this and this concerns SFB mode (for example with large virgin raw events), since it does not read input file in chunks in this mode.
FEDRawDataInputSource remains with the 32-bit limitation, but typically only around 200 MB buffers are used with it.

Parsifal-2045 committed:
Remove outdated parameters for "offline" Alpaka PixelTracks

cmsbuild committed:
Merge pull request #49306 from bsunanda/Phase2-hgx364L

Phase2-hgx364L Try to remove fine scintillator tile dependent tests for layers with no fine tiles

cmsbuild committed:
Merge pull request #49280 from adzel-werk/patch-1

Add InputTag parameters to fillDescriptions method for ctppsProtonReconstructionSimulationValidator_cfi

cmsbuild committed:
Merge pull request #49305 from alaperto/update_disabled_ROCs_DQM_Pixel_Summary_Master_16_0

Update Disabled ROCs for Online DQM Pixel Summary

smorovic committed:
add overflow check for chunk and block size and file chunkPosition_ and bufferPosition_

Commit:
disable ubsan for limited and one implementors

smuzaffar committed:
No need to explicitly delete the pre-existsing histogram for ROOT 6.38

Commit:
Avoid unnecessary printouts for normal execution

Commit:
Code check

cmsbuild committed:
Merge pull request #49303 from cms-tsg-storm/HLTdevelopment151X_for160X

HLT menu development for `15_1_X` (2/N) [`16_0_X`]

mmusich committed:
updated HLT menu in 2025 PbPb relval workflows

mmusich committed:
remove degeneracy PixelTrackProducerFromSoAAlpaka* and SiPixelRecHitFromSoAAlpaka*

mmusich committed:
clean-up HLT phase-2 menu from instances of PixelTrackProducerFromSoAAlpaka* and SiPixelRecHitFromSoAAlpaka*

pallabidas committed:
add EMTF NN hls model at setup

cmsbuild committed:
Merge pull request #49314 from Parsifal-2045/fixOfflinePixelTracksParameters

[NGT] Fix parameters for "offline" Phase 2 Alpaka PixelTracks

cmsbuild committed:
Merge pull request #49321 from bsunanda/Phase2-hgx364M

Phase2-hgx364M Avoid unnecessary printouts for normal execution

cmsbuild committed:
Merge pull request #49312 from smorovic/160x-buffer-size-fix

[DAQ] large input source buffer fix

alexandertuna committed:
Adding cluster size requirement to LST

mmusich committed:
demote 2024 HIon workflows to use @fake2

smuzaffar committed:
[PY312]PySys_SetArgv is deprecated in py3.12 and removed in py3.13

cmsbuild committed:
Merge pull request #49302 from alaperto/Cosmics_Strip_Clusters_DQM_Master_16_0

Cosmics Strip Clusters thresholds DQM

mmusich committed:
add Pixel rechits,tracks and vertices SoA monitoring for HLT heterogeneous workflows

smuzaffar committed:
DQMEcalMonitor: Root now requires bins in increasing order

cmsbuild committed:
Merge pull request #49322 from mmusich/mm_update_HLTmenu_2025PbPb_relvals

updated HLT menu in 2024 and 2025 PbPb relval workflows

mmusich committed:
fix typos and histogram titles in SiPixelCompareRecHits*

goblirsc committed:
modify MPS scripts for automated alignment

AuroraPerego committed:
add EM/Had energy in rechit to dumper

hjbossi committed:
Add uHTR data TP plots

Dr15Jones committed:
Use RNTuple streamer mode for more SoA classes

cmsbuild committed:
Merge pull request #49330 from smuzaffar/ecal-dqm-bin-order

DQMEcalMonitor: Root now requires bins in increasing order

mmusich committed:
add harvesting for HLTHeterogeneousMonitoring workflow

cmsbuild committed:
Merge pull request #49324 from mmusich/mm_cleanUpTrackeringAlpakaSpecialization

Clean-up specializations of `PixelTrackProducerFromSoAAlpaka` and `SiPixelRecHitFromSoAAlpaka`

cmsbuild committed:
Merge pull request #49326 from amecca/fix-extractPixBary

[TkAl][PixBary][FIX] Do not clobber output of multiple PixBary extract

hjbossi committed:
Add variables instead of hard coded values for TP info.

hjbossi committed:
Fix code checks.

guitargeek committed:
Don't use GLEW in Fireworks

The Frame Buffer Object (FBO) feature is part of OpenGL since version
3.0 from 2008. Nowadays, we can therefore assume that this feature
is always available. That means we don't have to use GLEW to check
whether this feature is present.

Like that, we can avoid using GLEW entirely, which is an implementation
detail of OpenGL loading in ROOT that will soon go away.

To make sure GLEW is not included anymore, this commit replaces includes
of `TGLIncludes.h` with the OpenGL headers only, because `TGLIncludes.h`
was including OpenGL plus GLEW.

See also:

  * https://github.com/root-project/root/issues/18471
  * https://github.com/root-project/root/pull/20294#issuecomment-3501449951

Moanwar committed:
Fix energy regression issue in TICLv5

cmsbuild committed:
Merge pull request #49340 from mmusich/mm_hlt_gpu_vs_cpu_comparisons_in_offline_dqm

Implement GPU vs CPU comparison for HLT pixel tracking heterogeneous products in patatrack workflows

martinamalberti committed:
fix for empty btl digi frames

cmsbuild committed:
Merge pull request #49338 from Dr15Jones/SOArntuple

Use RNTuple streamer mode for more SoA classes

cmsbuild committed:
Merge pull request #49319 from smuzaffar/fix-root639-histogram-delete

No need to explicitly delete the pre-existsing histogram for ROOT >= 6.38

cmsbuild committed:
Merge pull request #49323 from mmusich/HLT_HeterogeneousOnly

[NGT] Add a "Heterogeneous only" HLT phase 2 path and menu

cmsbuild committed:
Merge pull request #49325 from pallabidas/EMTF_setup_hls

[L1T] replacing tensorflow with hls4ml model for EMTF at setup

nothingface0 committed:
Fixes for personalPlayback.py

Add arguments for run_key, scram arch, fix path for cmsset_default,
fix log call, fix temporary file write mode

cmsbuild committed:
Merge pull request #49348 from nothingface0/fix_personal_playback

[DQM] Fixes for personalPlayback.py

cmsbuild committed:
Merge pull request #49337 from goblirsc/goblirsc-scriptsForAutoAlign

modify MPS scripts to support automated alignment

cmsbuild committed:
Merge pull request #49328 from alexandertuna/ClusterSizeRequirementRevived

Add cluster size requirement to LST

cmsbuild committed:
Merge pull request #49273 from hjbossi/FSCp_DQM_master

[FSC DQMn - master]: Initial commit of FSC plus implementation in DQM.

cmsbuild committed:
Merge pull request #49299 from Dr15Jones/matrixRNTupleTemp_master

Use RNTupleTemp* modules

smuzaffar committed:
Py312: use sys._getframe instead of slow inspect.getframeinfo

Commit:
rearrange comments per suggestion

Commit:
formatting

cmsbuild committed:
Merge pull request #49300 from smuzaffar/py312-deprecated-PySys_SetArgv

[PY312]PySys_SetArgv is deprecated in py3.11 and removed in py3.13

Commit:
Modify the scripts to add new scenarios in the testing

martinamalberti committed:
code-format

Dr15Jones committed:
Need to add TracksHost::Implementation to classes_def.xml

This avoids a segmentation fault in new version of ROOT.

martinamalberti committed:
tofDelay parameter removed, TDC window emulation implemented

cmsbuild committed:
Merge pull request #49343 from Moanwar/from-CMSSW_16_0_X_2025-11-06-2300

Fix energy regression issue in TICLv5

lwang046 committed:
Adjust led threshold for misfiremon

iarspider committed:
TrackerGeometry: add assert() to offsetDU and endsetDU

cmsbuild committed:
Merge pull request #49353 from AuroraPerego/hitsDumper

Add possibility to save rechits and simhits in the TICL Dumper

cmsbuild committed:
Merge pull request #49352 from smuzaffar/py312-use-getframe

Py312: use sys._getframe instead of slow inspect.getframeinfo

cmsbuild committed:
Merge pull request #49341 from guitargeek/glew

Don't use GLEW in Fireworks

perrotta committed:
Update the JEC tags in analysis GTs for 2022-24

martinamalberti committed:
fix TProfile ranges

mmusich committed:
move 2024 data workflows to run @standardDQMFakeHLT

Dr15Jones committed:
Fix UBSAN error in FWCore/Framework unit tests

- No longer dereference a nullptr
- Convert tests to catch2

cmsbuild committed:
Merge pull request #48629 from leobeltra/soa_blocks

Added SoABlocks feature

cmsbuild committed:
Merge pull request #49354 from perrotta/updateJecTagsInGts

Update the JEC tags in analysis GTs for 2022-24

cmsbuild committed:
Merge pull request #49364 from lwang046/HCALDQM-ledmon

HCALDQM: Adjust misfire threshold for LED calibration monitoring

cmsbuild committed:
Merge pull request #49369 from mmusich/move_2024_wfs_toRun_standardDQMFakeHLT

move 2024 data workflows to run `@standardDQMFakeHLT`

cmsbuild committed:
Merge pull request #49360 from Dr15Jones/fixTracksHostImplementation

Need to add TracksHost::Implementation to classes_def.xml

cmsbuild committed:
Merge pull request #49356 from bsunanda/Phase2-hgx364N

Phase2-hgx364N Modify the scripts to add new scenarios in the testing

cmsbuild committed:
Merge pull request #49357 from martinamalberti/mm_fixEmptyDigisBTL

MTD digitization: fix for empty BTL digis

cmsbuild committed:
Merge pull request #49190 from Electricks94/MultiCollection

Define RefProdVector for HGCRecHitCollection and PFRecHitCollection

cmsbuild committed:
Merge pull request #49361 from martinamalberti/mm_EtlTdcWindow

MTD digitization: update ETL digitization to emulate ETROC TDC window

soureek committed:
Updating fillDescription for BTV modules

soureek committed:
Merge branch 'cms-sw:master' into master

cmsbuild committed:
Merge pull request #49374 from fwyzard/Fix_BuildFiles_includes_and_comments

Fix BuildFiles, includes and comments

cmsbuild committed:
Merge pull request #49370 from Dr15Jones/ubsanFWFrameworkTestFix

Fix UBSAN error in FWCore/Framework unit tests

cmsbuild committed:
Merge pull request #49379 from soureek/master

Updating fillDescriptions in RecoBTag area

cmsbuild committed:
Merge pull request #49301 from dan131riley/ubsan-global-rsummary-bug

disable ubsan checks for some global summary functions

cmsbuild committed:
Merge pull request #49366 from iarspider/iarspider-patches-20251111-1

TrackerGeometry: add assert() to offsetDU and endsetDU
